app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt

```diff
@@ -210,14 +209,21 @@ public class MainActivity : AppCompatActivity(), MainViewController, RouteCallba
             true;
         })
         mapController?.setFeatureTouchListener({
-            properties ->
-                val tapProp = properties.getString(MAP_DATA_PROP_TAP)
-                val searchIndexProp = properties.getNumber(MAP_DATA_PROP_SEARCHINDEX).toInt()
-                if (tapProp == MAP_DATA_PROP_TAP_SEARCH) {
-                    presenter?.onSearchResultTapped(searchIndexProp)
+            properties, positionX, positionY ->
+                // Reassign tapPoint to center of the feature tapped
+                // Also used in placing the pin
+                poiTapPoint = floatArrayOf(positionX, positionY)
+                if (properties.contains(MAP_DATA_PROP_SEARCHINDEX)) {
+                    val searchIndex = properties.getNumber(MAP_DATA_PROP_SEARCHINDEX).toInt()
+                    presenter?.onSearchResultTapped(searchIndex)
                 } else {
-                    if (poiTapPoint != null) {
-                        presenter?.onReverseGeoRequested(poiTapPoint?.get(0) as Float, poiTapPoint?.get(1) as Float)
+                    if (properties.contains(MAP_DATA_PROP_ID)) {
+                        val featureID = properties.getNumber(MAP_DATA_PROP_ID).toLong()
+                        presenter?.onPlaceSearchRequested("osm:venue:$featureID")
+                    } else {
+                        if (poiTapPoint != null) {
+                            presenter?.onReverseGeoRequested(poiTapPoint?.get(0) as Float, poiTapPoint?.get(1) as Float)
```

Comment:
In this case I might have chosen to make the params type `Float?` and then do input validation in the presenter. And yes, in general `?.` and `toFloat()` are usually preferable to `!!.` or `as`.

There is also `as?` for null-safe casting but I haven't found a lot of places where this is useful so far.